### PR TITLE
[Feature]Add Confidence head and Dynamic Lengths to Verification Pipeline

### PR DESCRIPTION
**Pull request by StanislavII**
## Summary

This push adds confidence-based dynamic verification length scheduling for the DSpark speculative decoding path on Ascend.

The implementation uses the DSpark confidence head to estimate token survival probabilities and dynamically distribute a verification-token budget across active requests. Requests with higher predicted acceptance receive longer verification prefixes, while low-confidence requests verify fewer draft tokens.

### DSpark Confidence Head

- Added `DSparkConfidenceHead`.
- The confidence head combines:
  - the draft-model hidden state;
  - the corresponding Markov embedding.
- It produces a raw confidence logit for each speculative token.
- Confidence-head parameters are loaded separately from the base DSpark model parameters.
- Added validation to ensure that all expected confidence-head checkpoint tensors are loaded.

### Dynamic Verification Scheduling

- Added confidence and per-request verification-length buffers.
- Confidence logits are computed inside the autoregressive Markov-head drafting loop.
- Conditional confidence probabilities are converted into cumulative survival probabilities.
- A global verification budget is distributed across requests using the highest survival scores.
- Each request receives an individual verification length according to its predicted acceptance probability.

### Periodic Budget Selection

The global verification budget is recalculated once every `N` decoding steps instead of at every step.

This is based on the observation that the overall acceptance distribution usually changes gradually during generation. Recomputing the budget less frequently reduces scheduling overhead while still adapting to changes in draft quality.

The approximate budget-selection procedure is:

1. Apply `sigmoid` to the confidence logits.
2. Count confidence values above the configured threshold.
3. Compute the average selected length across active requests.
4. Round the result up to the next integer.
5. If the resulting value is even, increase it to the next odd integer.
6. Clamp the final budget to the supported speculative-token range.

Using odd budget values provides stable intermediate choices such as `1`, `3`, `5`, and `7`, avoiding unnecessary switching between neighboring verification lengths with similar execution costs.

### Minimum Verification Length

A configurable `min_k` guarantees that every active request receives at least a small verification prefix.

Without this lower bound, the global selector may assign zero tokens to low-confidence requests. These requests would make no speculative progress and could repeatedly remain in the low-confidence tail of the batch.

Enforcing `min_k >= 1` prevents this tail-stalling problem:

- every request continues to make progress;
- low-confidence requests are not permanently excluded;
- the remaining verification budget is still distributed according to survival probability.

The scheduler first assigns `min_k` tokens to every request and then distributes the remaining budget among the highest-scoring survival candidates.

### Per-Request Draft Truncation

- Generated draft blocks are truncated according to the selected verification length for each request.
- The scheduler receives only the selected draft-token prefix instead of the full static speculative block.
- This reduces unnecessary target-model verification for draft tokens that are unlikely to be accepted.

### Configuration

A new `dynamic_spec_config` section is added to `AscendConfig`.

```json
{
  "dynamic_spec_config": {
    "method": "dspark"
  }
}
````

The configuration is accessible through:

```python
get_ascend_config().dynamic_spec_config.method == "dspark"
```

When `method` is disabled (None), DSpark continues using the existing static verification behavior.

## Experimental Results

We evaluated dynamic verification scheduling on a mixed workload containing datasets with different DSpark acceptance profiles:

- **Low acceptance:** raw question answering
- **Medium acceptance:** agentic coding
- **High acceptance:** GSM8K

The experiments below used a batch size of `128` with `EAGER` mode

| Draft tokens | Dynamic scheduling | Throughput (tok/s) | Mean acceptance length | Throughput change |
|-------------:|:------------------:|-------------------:|-----------------------:|------------------:|
| 9 | OFF | 2812 | 4.27 | — |
| 9 | ON  | 3081 | 4.10 | **+9.56%** |
| 7 | OFF | 3022 | 4.16 | — |
| 7 | ON  | 3186 | 3.98 | **+5.43%** |
| 5 | OFF | 3149 | 3.68 | — |
| 5 | ON  | 3087 | 3.57 | **-2.00%** |

The experiments below used a batch size of `128` with `PIECEWISE` graph mode.

| Draft tokens | Dynamic scheduling | Throughput (tok/s) | Mean acceptance length | Throughput change |
|-------------:|:------------------:|-------------------:|-----------------------:|------------------:|
| 9 | OFF | 3052 | 4.26 | — |
| 9 | ON  | 3433 | 4.03 | **+12.48%** |
| 7 | OFF | 3347 | 4.16 | — |
| 7 | ON  | 3603 | 3.96 | **+7.65%** |
| 5 | OFF | 3602 | 3.70 | — |
| 5 | ON  | 3522 | 3.47 | **−2.22%** |

The experiments below used a batch size of `128` with `FULL_DECODE_ONLY` graph mode.

| Draft tokens | Dynamic scheduling | Throughput (tok/s) | Mean acceptance length | Throughput change |
|-------------:|:------------------:|-------------------:|-----------------------:|------------------:|
| 9 | OFF | 3147 | 4.26 | — |
| 9 | ON  | 3586 | 4.06 | **+13.95%** |
| 7 | OFF | 3452 | 4.16 | — |
| 7 | ON  | 3753 | 4.00 | **+8.72%** |
| 5 | OFF | 3690 | 3.70 | — |
| 5 | ON  | 3638 | 3.54 | **−1.41%** |

The experiments below used a batch size of `128` with `FULL` graph mode (number of testing prompts was reduced, so it's best not to use this table for comparison with other graph modes)

| Draft tokens | Dynamic scheduling | Throughput (tok/s) | Mean acceptance length | Throughput change |
|-------------:|:------------------:|-------------------:|-----------------------:|------------------:|
| 9 | OFF | 3133 | 4.27 | — |
| 9 | ON  | 3576 | 4.09 | **+14.14%** |
| 7 | OFF | 3428 | 4.16 | — |
| 7 | ON  | 3708 | 4.02 | **+8.17%** |
| 5 | OFF | 3639 | 3.70 | — |
| 5 | ON  | 3688 | 3.49 | **+0.13%** |
## Analysis

Dynamic scheduling is most beneficial when the maximum draft length is relatively large. For 9 draft tokens, it improves throughput by approximately 12–14%, and for 7 draft tokens by 8–9%, despite slightly reducing the mean acceptance length. This indicates that the reduction in target verification work outweighs the loss in accepted tokens per step. At 5 draft tokens, there is little remaining verification work to remove, so the scheduling overhead and lower acceptance length largely cancel the benefit. 

Dynamic scheduling also makes performance substantially less sensitive to the configured maximum draft length. Without scheduling, throughput decreases considerably as the draft length grows from 5 to 9. With scheduling enabled, the throughput values are much closer because requests that do not benefit from long verification lengths are truncated dynamically.
## Goal

The goal of this change is to reduce unnecessary target-model verification work while preserving useful draft tokens.

High-confidence requests can use longer verification blocks, while low-confidence requests use shorter prefixes. Periodic budget recalculation limits scheduling overhead, and `min_k` ensures that all requests continue making progress.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/0351e9aa1fdf1a51329d1906881528dfe61fc88e

### CODE REVIEW

File: tests/e2e/pull_request/one_card/spec_decode/test_dynamic.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            "method": "dspark",
+            "method_params": {
+                "initial_verify_budget_per_req": 3,
+                "budget_update_interval": 1,
+                "budget_threshold": 0.7,
+            },
+        },
+    }
+
+    compilation_config = CompilationConfig(cudagraph_mode="PIECEWISE", cudagraph_capture_sizes=[7, 8])
```
Comment (drslark) [inline tests/e2e/pull_request/one_card/spec_decode/test_dynamic.py]: Can you support target `FULL` with draft `eager`?

File: vllm_ascend/spec_decode/dspark_proposer.py [hunk 2]
```diff
@@ -102,6 +138,109 @@ def __init__(
...
+
+    def _compute_verify_budget(self, confidence_logits: torch.Tensor) -> None:
+        """Recompute the per-request verify budget every `budget_update_interval` steps."""
+        self._steps_since_budget_update += 1
+        if self._steps_since_budget_update < self.budget_update_interval:
+            return
+        self._steps_since_budget_update = 0
+        num_reqs = confidence_logits.shape[0]
+        # Approximated budget allocation via averaged per-position anticipated acceptance.
+        mean_k = float((confidence_logits.sigmoid() > self.budget_threshold).sum().item()) / float(num_reqs)
```
Comment (drslark) [inline vllm_ascend/spec_decode/dspark_proposer.py:187]: Add a comment about synchronous event that `item()` introduces.

**Review by drslark — CHANGES_REQUESTED** [review]

**Comment by drslark** [comment]
Please add a doc that only qwen3-8b supports dynamic lengths.
And it is an experimental feature, which will be supported only in `v2`.

**Comment by chenaoxuan** [comment]
> Please add a doc that only qwen3-8b supports dynamic lengths. And it is an experimental feature, which will be supported only in `v2`.

No, not only Qwen3-8B, every models with Confidence-Head will support our dynamic method.

**Comment by drslark** [comment]
> > Please add a doc that only qwen3-8b supports dynamic lengths. And it is an experimental feature, which will be supported only in `v2`.
> 
> No, not only Qwen3-8B, every models with Confidence-Head will support our dynamic method.

But only with PIECEWISE.

File: vllm_ascend/spec_decode/dspark_proposer.py (hunk 2 above)
Comment (StanislavII) [reply vllm_ascend/spec_decode/dspark_proposer.py:187]: .item() waits for the NPU computation to finish and copies the result to the CPU, so this introduces a synchronization on only budget-update steps

Comment (drslark) [reply vllm_ascend/spec_decode/dspark_proposer.py:187]: I know it.
So we need a comment or TODO to mark it.

File: tests/e2e/pull_request/one_card/spec_decode/test_dynamic.py (hunk 1 above)
Comment (StanislavII) [reply tests/e2e/pull_request/one_card/spec_decode/test_dynamic.py]: yes, it supports - "Experimental Results" section above

**Review by weijinqian0 — APPROVED** [review]

File: vllm_ascend/ascend_config.py [hunk 3]
```diff
@@ -46,6 +46,9 @@ def __init__(self, vllm_config: "VllmConfig"):
         finegrained_tp_config = additional_config.get("finegrained_tp_config", {})
         self.finegrained_tp_config = FinegrainedTPConfig(finegrained_tp_config, vllm_config)
 
+        dynamic_spec_config = additional_config.get("dynamic_spec_config", {})
```
Comment (kunpengW-code) [inline vllm_ascend/ascend_config.py:49]: Added the parameter description of the `additional_config.md` file.

Comment (chenaoxuan) [reply vllm_ascend/ascend_config.py:49]: Done.

File: docs/source/user_guide/feature_guide/speculative_decoding.md [hunk 4]
```diff
@@ -453,73 +453,98 @@ Key configuration parameters:
...
 
-### Block Verify
+### Confidence-head based (DSpark)
 
-Block Verify evaluates all draft tokens as a block using cumulative probability products, rather than checking each token independently. This can improve the acceptance rate and reduce the overhead of rejection sampling, especially when `num_speculative_tokens >= 3`.
+This approach adapts how many drafted tokens are verified per request. It can reduce verify cost when the drafter is less confident about later tokens, while still allowing longer speculation when confidence is high.
 
-### Entropy Verify
+> [!NOTE]
+> This is an **exploratory** feature and currently targets **model runner v1** only. The supported dynamic method relies on the DSpark **confidence head**. For other proposers such as DFlash, non-neural-parameter-based dynamic schemes are under exploration. Among DSpark draft models, only the **Qwen** series is supported today; support for models such as GLM and DeepSeek is being added incrementally.
```
Comment (drslark) [inline docs/source/user_guide/feature_guide/speculative_decoding.md:468]: Good.
Very clear.

**Review by drslark — APPROVED** [review]
LGTM.
Thanks for your efforts.

File: docs/source/user_guide/feature_guide/speculative_decoding.md (hunk 4 above)
Comment (drslark) [reply docs/source/user_guide/feature_guide/speculative_decoding.md:468]: I'm wondering why `FULL_DECODE_ONLY` works.
I'll check it myself later.
